mautic/mautic
# Truly transactional emails

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [❌]
| New feature/enhancement? (use the a.x branch)      | [✅]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [✅]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      |  
| Related developer documentation PR URL |  
| Issue(s) addressed                     | https://acquia.atlassian.net/browse/MAUT-11283 

#### Description:

This PR is adding a new option for all segment and template emails to force-send the email to contacts who have unsubscribed. This enables marketers to send truly-transactional emails. But it works for marketing emails the same. The use case may be to send updated terms and conditions email to all contacts. Here's the new switch:
![Screenshot 2024-08-27 at 09 43 28](https://github.com/user-attachments/assets/d5ab9a44-a097-4655-bc00-2e34807a8a42)

Since this can be a legally dangerous setting, there is an extra warning message that the users must confirm to send the email also to users who have unsubscribed:
![Screenshot 2024-08-27 at 17 07 20](https://github.com/user-attachments/assets/0b6620f1-09a1-41c3-abc5-86925e6154aa)

This PR also changes the way the frequency rules are calculated and applied. The frequency rules will only consider the “Send to unsubscribed contacts” toggle. If the toggle is set to NO, such emails will be counted towards the calculation of Frequency Rule. If the toggle is YES, such emails will not be counted towards the calculation of Frequency Rules.

The only place where to configure the "transactional vs marketing" emails was in the Send Email campaign action. This is how the email send action looked like before:
![Screenshot 2024-08-27 at 17 05 43](https://github.com/user-attachments/assets/2ad2cf55-afcb-4a32-9b38-5d815953399d)

And if the "Marketing" option was chosen then the send email action looked like this:
![Screenshot 2024-08-27 at 17 05 50](https://github.com/user-attachments/assets/31af3d70-502a-4ccf-b04a-42e96da3f2f8)

Now that witch looks like this with the tooltip:
![Screenshot 2024-08-27 at 17 08 25](https://github.com/user-attachments/assets/6c2235bb-1212-4d43-9695-8e7a9a77c8b0)

This switch is still there, but renamed from *Email Type: Transactional / Marketing* to *Repeatable email: Yes/No* which better describes the function. This is how it looks like before an email is selected:
![Screenshot 2024-08-27 at 17 07 47](https://github.com/user-attachments/assets/fb308d90-468b-4cf9-85a5-d6c168209e89)

This is how it looks like when a normal email is selected:
![Screenshot 2024-08-27 at 17 08 03](https://github.com/user-attachments/assets/7861d351-9e21-4a5d-95d3-ab23f31caa85)

And this is how it looks like when an email that ignores DNC is selected:
![Screenshot 2024-08-27 at 17 08 15](https://github.com/user-attachments/assets/4f72a770-8efd-42d5-a995-074a35df78d5)

There is also a new global config option whether the unsubscribe headers should be added to emails which obey the "Do Not Contact" (DNC) option:
![Screenshot 2024-08-27 at 17 00 21](https://github.com/user-attachments/assets/3abc71f9-7c3a-4cb4-b54a-f4ac3fc1267b)

This PR also contains the fix for sending multiple emails per API request from https://github.com/mautic/mautic/pull/13998

Another change in this PR is refactoring the `MailHelper` finally doesn't use the deprecated `MauticFactory` but explicitly defines its dependencies via DI.

#### Steps to test this PR:

Please test this with SMTP as well as a API based transport. This [Sparkpost PR](https://github.com/acquia/mc-cs-plugin-sparkpost/pull/3) is ready to accommodate the changes in this PR.

2. Verify new role permission "Include unsubscribed contacts" inside email permission
3. Verify "Include unsubscribed contacts" toggle in Email advance tab for both type of email (Template/Segment)
4. Verify new toggle in email advance tab should be editable based on new role permission
5. Verify New toggle value should visible in Campaign Action/Point Trigger/Form action on email selection.
6. Send email from campaign, "Include unsubscribed contacts" should work if yes then send email to all contacts including DNC

1. Create a campaign that uses the "Send Email" event. 
2. Choose "Transactional" as your email type.
3. You should not get emails with an Unsubscribe link.

1. Make sure the `Do not send me more than` is set to `0` (means disabled) under the `Configuration -> Email Settings -> Default Frequency Rule`. 
    ![Screenshot 2024-03-15 at 14 51 52](https://github.com/acquia/mc-cs/assets/6388925/186eab71-72d7-43d0-957b-53533b404a54)
2. Create a new contact.
3. On the contact's detail page click the `Preferences` and set the email frequency to `2` a `month` and click "Save"
    ![Screenshot 2024-03-15 at 14 50 55](https://github.com/acquia/mc-cs/assets/6388925/00941d1b-7b60-4414-833c-8dfa714a6101)
4. Send 3 different emails with “Send to unsubscribed contacts” toggle set to NO to the contact. Only first two emails should be sent immediately. The third one should be postponed by a month.
5. Send 2 different emails with “Send to unsubscribed contacts” toggle set to YES to the contact. Both emails should be sent immediately.
6. Set `Do not send me more than` is set to `2` Messages each `month` under the `Configuration -> Email Settings -> Default Frequency Rule`. 
7. Create a new contact.
8. Send 3 different emails with “Send to unsubscribed contacts” toggle set to NO to the contact. Only first two emails should be sent immediately. The third one should be postponed by a month.
9. Send 2 different emails with “Send to unsubscribed contacts” toggle set to YES to the contact. Both emails should be sent immediately.
2. Verify new added toggle `Disable unsubscribe link` in email setting
> Unsubscribe header should be automatically include only if toggle is off
3. Verify one click unsubscribe header to all email where `send to unsubscribe list` is off 

#### List deprecations along with the new alternative:
1. Previous Marketing/Transactional flag is no longer considered for applying frequency rules. 

[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. Applying the frequency rules.
2. Counting emails towards the frequency rules.
